src/com/google/javascript/jscomp/ConvertChunksToESModules.java

```diff
@@ -0,0 +1,297 @@
+      JSModule importingChunk = chunkImportsEntry.getKey();
+      CompilerInput firstInput = importingChunk.getInput(0);
+      Node moduleBody = firstInput.getAstRoot(compiler).getFirstChild();
+      checkState(moduleBody != null && moduleBody.isModuleBody());
+
+      // For each distinct chunk where a referenced symbol is defined, create an import statement
+      // referencing the names.
+      for (Map.Entry<JSModule, Set<String>> importsByChunk : chunkImportsEntry.getValue().entrySet()) {
+        Node importSpecs = new Node(Token.IMPORT_SPECS);
+        for (String name : importsByChunk.getValue()) {
+          Node importSpec = new Node(Token.IMPORT_SPEC);
+          importSpec.addChildToFront(IR.name(name));
+          importSpec.addChildToFront(IR.name(name));
+          importSpec.putIntProp(Node.IS_SHORTHAND_PROPERTY, 1);
+          importSpecs.addChildToBack(importSpec);
+        }
+        Node importStatement = new Node(Token.IMPORT);
+        JSModule exportingChunk = importsByChunk.getKey();
+        String importPath = GetChunkName(exportingChunk);
+        try {
+          importPath =
+              this.relativePath(
+                  GetChunkName(importingChunk),
+                  GetChunkName(exportingChunk));
+        } catch (IllegalArgumentException e) {
+          compiler.report(
+              JSError.make(
+                  moduleBody,
+                  UNABLE_TO_COMPUTE_RELATIVE_PATH,
+                  GetChunkName(importingChunk),
+                  GetChunkName(exportingChunk)));
+        }
+        importStatement.addChildToFront(IR.string(importPath));
+        importStatement.addChildToFront(importSpecs);
+        importStatement.addChildToFront(IR.empty());
+        importStatement.useSourceInfoFromForTree(moduleBody);
+        importStatements.add(0, importStatement);
+      }
+      for (Node importStatement : importStatements) {
+        moduleBody.addChildToFront(importStatement);
+      }
+      compiler.reportChangeToEnclosingScope(moduleBody);
+    }
+  }
+
+  /** Find names in a module that are defined in a different module. */
+  private class FindCrossChunkReferences extends AbstractPostOrderCallback {
+    @Override
+    public void visit(NodeTraversal t, Node n, Node parent) {
+      if (n.isName()) {
+        String name = n.getString();
+        if ("".equals(name)) {
+          return;
+        }
+        Scope s = t.getScope();
+        Var v = s.getVar(name);
+        if (v == null || !v.isGlobal()) {
+          return;
+        }
+        CompilerInput input = v.getInput();
+        if (input == null) {
+          return;
+        }
+        // Compare the chunk where the variable is declared to the current
+        // chunk. If they are different, the variable is used across modules.
+        JSModule definingChunk = input.getModule();
+        JSModule referencingChunk = t.getModule();
+        if (definingChunk != referencingChunk) {
+          if (NodeUtil.isLhsOfAssign(n)) {
```

Comment:
This will miss some cases. I believe you want `NodeUtil.isLValue(n)`.

Reply:
Fixed